vault/activity/query.go

```diff
@@ -196,42 +220,45 @@ func (s *PrecomputedQueryStore) Get(ctx context.Context, startTime, endTime time
 	sort.Slice(filteredList, func(i, j int) bool {
 		return filteredList[i].After(filteredList[j])
 	})
-	contiguous := timeutil.GetMostRecentContiguousMonths(filteredList)
-	actualStartTime := contiguous[len(contiguous)-1]
 
-	s.logger.Trace("chose start time", "actualStartTime", actualStartTime, "contiguous", contiguous)
-
-	endTimes, err := s.listEndTimes(ctx, actualStartTime)
-	if err != nil {
-		return nil, err
-	}
-	s.logger.Trace("retrieved end times from storage", "endTimes", endTimes)
-
-	// Might happen if there's a race with GC
-	if len(endTimes) == 0 {
-		s.logger.Warn("missing end times", "start time", actualStartTime)
-		return nil, nil
-	}
-	var actualEndTime time.Time
-	for _, t := range endTimes {
-		if timeutil.InRange(t, startTime, endTime) {
-			if actualEndTime.IsZero() || t.After(actualEndTime) {
-				actualEndTime = t
-			}
+	closestStartTime := time.Time{}
+	closestEndTime := time.Time{}
+	maxTimeDifference := time.Duration(0)
+	for i := len(filteredList) - 1; i >= 0; i-- {
+		testStartTime := filteredList[i]
+		s.logger.Trace("trying test start times", "startTime", testStartTime, "filteredList", filteredList)
+		testEndTime, err := s.getMaxEndTime(ctx, testStartTime, endTime)
+		if err != nil {
+			return nil, err
+		}
+		if testEndTime.IsZero() {
+			// Might happen if there's a race with GC
+			s.logger.Warn("missing end times", "start time", testStartTime)
+			continue
+		}
+		s.logger.Trace("retrieved max end time from storage", "endTime", testEndTime)
+		diff := testEndTime.Sub(testStartTime)
+		if diff >= maxTimeDifference {
+			closestStartTime = testStartTime
+			closestEndTime = testEndTime
+			maxTimeDifference = diff
+			s.logger.Trace("updating closest times")
```

Comment:
I might be breading this logic incorrectly but it does not necessarily seem like it aligns with the following:

> Changes the query get logic from an algorithm to find the most recent contiguous chunk of segments to finding the closest sub-interval to the user-specified query interval.

It seems like instead what is calculated is the _widest_ contiguous chunk. That may not necessarily be the most recent, correct?

Reply:
That's the same as what I said in the above quote, right? The "widest" contiguous chunk within the user-specified start and end time is the closest sub-interval to what the user specified.

Reply:
I was focusing on the "closest sub-interval to the user-specified query interval" part. To provide an example, say the user specified Jan-August and there are the following contiguous chunks, Jan-Feb, Feb-Mar, and Mar-May. Based on the logic, I _think_ that Mar-May would be selected. Maybe that's the expectation and I'm misunderstanding the intention of the change and if so, sorry for the confusion! I interpreted "closest sub-interval" as being either close to the start time or end time.

Reply:
The sub-interval isn't contiguous. So in the example you gave, all 3 intervals Jan-Feb, Feb-Mar, and Mar-May, should be returned with skipped nil months. 

Now, if the time interval specified was over 12 months (say we queried for Jan 2021 - April 2022), and the largest contiguous chunk of data is over 12 months (Jan 2021 - Feb 2022, say), then we will just return that contiguous chunk. The reason for this is that we simply need an upper bound for the maximum non-contiguous precomputation, so I chose a year.

Reply:
We met directly so that I could get a little more insight into this discussion. I initially did not account for the impact that the change from using `getMostRecentActivityLogSegment` to `getMostRecentNonContiguousActivityLogSegments` in `precomputedQueryWorker` had on `Get`. That discussion paired with the example above resolved my confusion. Thanks, @HridoyRoy!

Reply:
No problem, thanks for digging into it!